### PR TITLE
Bug Fix: Date spine timestamp error

### PR DESCRIPTION
**Pull request by fivetran-avinash**
## PR Overview
**This PR will address the following Issue/Feature:** [#23]

**This PR will result in the following new package version:** 0.5.0

No fields were changed in this update, just the behavior of the spine for empty general ledger models. 

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

# Bug Fixes
- Updated the `int_sage_intacct__general_ledger_date_spine` model to accommodate for the cases when the compiled `sage_intacct__general_ledger` model has no transactions. In this case, the model now defaults to a range of one-month from the current date.

## Under The Hood
- Updated the structure of the `int_sage_intacct__general_ledger_date_spine` model for improved performance and maintainability.
- Added `flags.WHICH in ('run', 'build')` as a condition in `int_sage_intacct__general_ledger_date_spine` to prevent call statements from querying the staging models during a `dbt compile`.
- Addition of integrity and consistency validation tests within integration tests for the `sage_intacct__general_ledger` and `sage_intacct__general_ledger_by_period` model.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [NA] dbt run (if incremental models are present)

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned.
- [NA] All necessary documentation and version upgrades have been applied.
    
- [x] docs were regenerated (unless this PR does not include any code or yml updates).
- [x] BuildKite integration tests are passing.
- [x] Detailed validation steps have been provided below.

### Detailed Validation
Please share any and all of your validation steps:

To reproduce the error, I created empty versions of the seed files that flowed into the general ledger model, in this case `gl_account`, `gl_batch` and `gl_detail`, then ran it in snowflake. 

Screenshot 2024-07-29 at 8 59 46 PM

Adding the new logic fixed the error on a `dbt run`. 

Also updating the date spine model to match our recent salesforce updates led to a successful `dbt compile`.
Screenshot 2024-07-26 at 2 20 39 PM

To validate these changes on devprod, I created:

- An integrity test that compared row counts to the `general_ledger_by_period` model that flows downstream from the new `general_ledger_date_spine`.
- I created a consistency test for the `general_ledger` to compare values between dev and prod. 
Screenshot 2024-07-29 at 8 04 19 PM

One note: 
- I originally created a consistency test on the `general_ledger_by_period` as well, but this test was failing. It appears the old logic uses the current date as the last date. So prod takes a lot of extra records that probably shouldn't exist since the max(entry_date_at) that created the max date on prod and dev is  almost a year old. 
 
Screenshot 2024-07-30 at 12 29 06 AM

So this test should fail on this PR, but should work for all future PRs. I've left it in here with the expectation that it'll fail here but is a useful test for future models. 

### If you had to summarize this PR in an emoji, which would it be?
 
🦴

### CODE REVIEW

File: CHANGELOG.md
Comment (fivetran-reneeli) [inline CHANGELOG.md:5]:
```suggestion
- Updated the `int_sage_intacct__general_ledger_date_spine` model to accommodate for the cases when the compiled `sage_intacct__general_ledger` model has no transactions. In this case, the model now defaults to a range of one month leading up to the current date.
```

**Review by fivetran-reneeli — COMMENTED** [review]
Looks good just to set it as a breaking change!

**Review by fivetran-reneeli — APPROVED** [review]
lgtm

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash thanks for your work on this PR. It's looking great, I just have a few small notes before approval. Let me know if you want to sync on any of my notes.

File: integration_tests/tests/integrity/integrity_general_ledger_by_period.sql
Comment (fivetran-avinash) [reply integration_tests/tests/integrity/integrity_general_ledger_by_period.sql:16]: @fivetran-joemarkiewicz Makes sense. Just checking--this last line is supposed to be the `sage_intacct__general_ledger_by_period`?

Comment (fivetran-joemarkiewicz) [reply integration_tests/tests/integrity/integrity_general_ledger_by_period.sql:16]: Ah yes, copy/paste error on my part

Comment (fivetran-avinash) [reply integration_tests/tests/integrity/integrity_general_ledger_by_period.sql]: Updated

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md]: Updated

**Comment by fivetran-avinash** [comment]
> @fivetran-avinash thanks for your work on this PR. It's looking great, I just have a few small notes before approval. Let me know if you want to sync on any of my notes.

@fivetran-joemarkiewicz This is ready for re-review.

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM

**Pull request merged**
